This notebook is for the very initial preprocessing of our chosen data.
Author: loua

Choices made regarding missing data:
Since we are dealing with just one data type - that is, the number of people (integers) - we do the following: If a column has some cells with data > 0, then we set cells with missing data in that column to 0. If a column only has rows with missing data, we set all 0/'-' to NaN and drop the column.

#### Done
- socioeconomy
- citizenship (and age and sex)
- education
- immigrants
- tenure
- type
- simple (contains income, car, benefit type, arrests)
- builtyear (housing construction year)

#### Extras
- [] Add valgsted navn(e), kredsnavn
- [] Area in sq metres
- [] Put election data in same format (one csv for each year)

#### TO DO
- [] Citizenship but only for 20+ years

In [1]:
# Libraries
import pandas as pd
import combine_raw_data as cb

# Configuration
from combine_raw_data import ROOT, PARENT_PATH, PATH, OUTPUT_PATH, YEARS, MUNICIPALITY_IDS, COMMON_COLS, PREFIX, CPH_AREAS


## PersonsBySocioeconomyIndustry

We want data for arbejdsløse, topledere and selvstændige. But there are similar column names for both topledere and selvstændige - we need to check these columns to see if data is duplicated, different, or only exists in one of the similarly named columns. 

In [2]:
# Socio-økonomisk status og brancher fordelt på afstemningsområder,Population_PersonsBySocioeconomyIndustry.csv
path = PATH + 'PersonsBySocioeconomyIndustry.csv'
prefixes = [f"{PREFIX}{year}" for year in YEARS]
pattern = ' - Socio-økonomisk status og brancher fordelt på afstemningsområder_'
wanted_cols = [
    '07. Arbejdsløse_',
    '02. Topledere_',
    '01. Selvstændig og medhj._'
]
cols = []

df = pd.read_csv(path, sep=';', dtype=str)
df = df[df['Gruppe'].isin(CPH_AREAS)]

for prefix in prefixes:
    for name in wanted_cols:
        colname_start = f'{prefix}{pattern}{name}'
        for col in df.columns:
            if col.startswith(colname_start):
                cols.append(col)

final_cols = COMMON_COLS + cols
df = df[final_cols]

df_transposed = df.transpose()

df_transposed.tail()

,0,1,2,3,4,5,6,7,8,9,...,48,49,50,51,52,53,54,55,56,57
KV2021 - Socio-økonomisk status og brancher fordelt på afstemningsområder_01. Selvstændig og medhj._Information og kommunikation,74,43,76,47,46,46,129,54,119,57,...,47,51,19,69,62,63,41,39,28,77
KV2021 - Socio-økonomisk status og brancher fordelt på afstemningsområder_01. Selvstændig og medhj._Kultur fritid og anden service,73,41,88,62,48,56,158,78,156,50,...,40,89,35,68,57,72,62,47,45,112
KV2021 - Socio-økonomisk status og brancher fordelt på afstemningsområder_01. Selvstændig og medhj._Landbrug skovbrug og fiskeri,3,4,5,-,2,2,5,-,9,7,...,4,1,2,5,1,1,3,1,4,2
KV2021 - Socio-økonomisk status og brancher fordelt på afstemningsområder_01. Selvstændig og medhj._Offentlig administration undervisning og sundhed,95,50,100,60,76,49,157,53,133,82,...,73,27,24,84,92,103,70,66,67,144
KV2021 - Socio-økonomisk status og brancher fordelt på afstemningsområder_01. Selvstændig og medhj._Uoplyst aktivitet,-,-,-,-,-,-,-,-,-,-,...,-,-,-,-,-,-,-,-,-,-


In [3]:
cb.inspect_socioeconomy_duplicates(2001, 20) # don't use any - no data at all
cb.inspect_socioeconomy_duplicates(2005, 20) # don't use any - no data at all
cb.inspect_socioeconomy_duplicates(2009, 20) # use cols with no number (numbered cols don't have data)
cb.inspect_socioeconomy_duplicates(2013, 20) # use cols with no number (numbered cols don't have data)
cb.inspect_socioeconomy_duplicates(2017, 20) # use cols with no number (numbered cols don't have data)
cb.inspect_socioeconomy_duplicates(2021, 20) # use either one - both have data - data is different, but with high correlation for most columns



SOCIOECONOMY DUPLICATE INSPECTION - Year 2001


01. Selvstændig og medhj.:
  Found 23 columns
  Industries with duplicates: 11
  Industries with single column: 1

  Analyzing 11 duplicate pairs:

  Industry: Landbrug skovbrug og fiskeri
    Version 1 (numbered):
      Non-null: 58/58, Non-zero: 58, Sum: 0.00, Mean: nan
      Sample values: [nan, nan, nan]
    Version 2 (no number):
      Non-null: 58/58, Non-zero: 58, Sum: 0.00, Mean: nan
      Sample values: [nan, nan, nan]
    Comparison:
      Identical: True
      Correlation: Cannot calculate (no variation in data)
      Sum difference: 0.00

  Industry: Industri råstofindvinding og forsyningsvirksomhed
    Version 1 (numbered):
      Non-null: 58/58, Non-zero: 58, Sum: 0.00, Mean: nan
      Sample values: [nan, nan, nan]
    Version 2 (no number):
      Non-null: 58/58, Non-zero: 58, Sum: 0.00, Mean: nan
      Sample values: [nan, nan, nan]
    Comparison:
      Identical: True
      Correlation: Cannot calculate (no variation i

{'01. Selvstændig og medhj.': ['KV2021 - Socio-økonomisk status og brancher fordelt på afstemningsområder_01. Selvstændig og medhj._01. Landbrug skovbrug og fiskeri',
  'KV2021 - Socio-økonomisk status og brancher fordelt på afstemningsområder_01. Selvstændig og medhj._02. Industri råstofindvinding og forsyningsvirksomhed',
  'KV2021 - Socio-økonomisk status og brancher fordelt på afstemningsområder_01. Selvstændig og medhj._03. Bygge og anlæg',
  'KV2021 - Socio-økonomisk status og brancher fordelt på afstemningsområder_01. Selvstændig og medhj._04. Handel og transport mv.',
  'KV2021 - Socio-økonomisk status og brancher fordelt på afstemningsområder_01. Selvstændig og medhj._05. Information og kommunikation',
  'KV2021 - Socio-økonomisk status og brancher fordelt på afstemningsområder_01. Selvstændig og medhj._06. Finansiering og forsikring',
  'KV2021 - Socio-økonomisk status og brancher fordelt på afstemningsområder_01. Selvstændig og medhj._07. Ejendomshandel og udlejning',
  'KV2

In [4]:
socio_2001 = cb.preprocess_socioeconomy(2001)
socio_2005 = cb.preprocess_socioeconomy(2005)
socio_2009 = cb.preprocess_socioeconomy(2009)
socio_2013 = cb.preprocess_socioeconomy(2013)
socio_2017 = cb.preprocess_socioeconomy(2017)
socio_2021 = cb.preprocess_socioeconomy(2021)

2001
    01. Selvstændig og medhj.: Using 12 unnumbered columns
    02. Topledere: Using 12 unnumbered columns
    07. Arbejdsløse: Using 12 unnumbered columns
2005
    01. Selvstændig og medhj.: Using 12 unnumbered columns
    02. Topledere: Using 12 unnumbered columns
    07. Arbejdsløse: Using 12 unnumbered columns
2009
    01. Selvstændig og medhj.: Using 12 unnumbered columns
    02. Topledere: Using 12 unnumbered columns
    07. Arbejdsløse: Using 12 unnumbered columns
2013
    01. Selvstændig og medhj.: Using 12 unnumbered columns
    02. Topledere: Using 12 unnumbered columns
    07. Arbejdsløse: Using 12 unnumbered columns
2017
    01. Selvstændig og medhj.: Using 12 unnumbered columns
    02. Topledere: Using 12 unnumbered columns
    07. Arbejdsløse: Using 12 unnumbered columns
2021
    01. Selvstændig og medhj.: Using 12 unnumbered columns
    02. Topledere: Using 12 unnumbered columns
    07. Arbejdsløse: Using 12 unnumbered columns


In [5]:
print(f"{'='*80}\n  2001\n")
print(socio_2001.info())
print(f"\n{'='*80}\n  2005\n")
print(socio_2005.info())
print(f"\n{'='*80}\n  2009\n")
print(socio_2009.info())
print(f"\n{'='*80}\n  2013\n")
print(socio_2013.info())
print(f"\n{'='*80}\n  2017\n")
print(socio_2017.info())
print(f"\n{'='*80}\n  2021\n")
print(socio_2021.info())

  2001

<class 'pandas.core.frame.DataFrame'>
Index: 58 entries, 0 to 57
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Gruppe                      58 non-null     object 
 1   ValgstedId                  58 non-null     object 
 2   KredsNr                     58 non-null     object 
 3   StorKredsNr                 58 non-null     object 
 4   LandsdelsNr                 58 non-null     object 
 5   Socioeconomy_Selvstaendige  0 non-null      float64
 6   Socioeconomy_Topledere      0 non-null      float64
 7   Socioeconomy_Arbejdsloese   0 non-null      float64
 8   Socioeconomy_Total          58 non-null     float64
dtypes: float64(4), object(5)
memory usage: 4.5+ KB
None

  2005

<class 'pandas.core.frame.DataFrame'>
Index: 58 entries, 0 to 57
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      -------------- 

In [6]:
socio_2021.head()

,Gruppe,ValgstedId,KredsNr,StorKredsNr,LandsdelsNr,Socioeconomy_Selvstaendige,Socioeconomy_Topledere,Socioeconomy_Arbejdsloese,Socioeconomy_Total
0,101001,101001,1,1,1,563.0,383.0,305.0,1251.0
1,101002,101002,1,1,1,357.0,221.0,137.0,715.0
2,101003,101003,1,1,1,632.0,401.0,277.0,1310.0
3,101005,101005,1,1,1,396.0,230.0,312.0,938.0
4,101006,101006,1,1,1,410.0,246.0,302.0,958.0


The total calculated before is wrong - need to total across all columns in the csv, incl the ones we're not using. Using a different function:

In [7]:
socio_2001 = cb.preprocess_socioeconomy2(2001)
socio_2005 = cb.preprocess_socioeconomy2(2005)
socio_2009 = cb.preprocess_socioeconomy2(2009)
socio_2013 = cb.preprocess_socioeconomy2(2013)
socio_2017 = cb.preprocess_socioeconomy2(2017)
socio_2021 = cb.preprocess_socioeconomy2(2021)

print(f"{'='*80}\n  2001\n")
print(socio_2001.info())
print(f"\n{'='*80}\n  2005\n")
print(socio_2005.info())
print(f"\n{'='*80}\n  2009\n")
print(socio_2009.info())
print(f"\n{'='*80}\n  2013\n")
print(socio_2013.info())
print(f"\n{'='*80}\n  2017\n")
print(socio_2017.info())
print(f"\n{'='*80}\n  2021\n")
print(socio_2021.info())

2001
2005
2009
2013
2017
2021
  2001

<class 'pandas.core.frame.DataFrame'>
Index: 58 entries, 0 to 57
Data columns (total 9 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Gruppe                       58 non-null     object 
 1   ValgstedId                   58 non-null     object 
 2   KredsNr                      58 non-null     object 
 3   StorKredsNr                  58 non-null     object 
 4   LandsdelsNr                  58 non-null     object 
 5   Socioeconomy_Self-employed   58 non-null     float64
 6   Socioeconomy_Top executives  58 non-null     float64
 7   Socioeconomy_Unemployed      58 non-null     float64
 8   Socioeconomy_Total           58 non-null     float64
dtypes: float64(4), object(5)
memory usage: 4.5+ KB
None

  2005

<class 'pandas.core.frame.DataFrame'>
Index: 58 entries, 0 to 57
Data columns (total 9 columns):
 #   Column                       Non-Null Count  Dtype  
---  -

In [8]:
socio_2021.head()

,Gruppe,ValgstedId,KredsNr,StorKredsNr,LandsdelsNr,Socioeconomy_Self-employed,Socioeconomy_Top executives,Socioeconomy_Unemployed,Socioeconomy_Total
0,101001,101001,1,1,1,563.0,383.0,305.0,15540.0
1,101002,101002,1,1,1,357.0,221.0,137.0,8366.0
2,101003,101003,1,1,1,632.0,401.0,277.0,15838.0
3,101005,101005,1,1,1,396.0,230.0,312.0,15817.0
4,101006,101006,1,1,1,410.0,246.0,302.0,13473.0


### Conclusion

There are no data for 2001 and 2005, but there are data for the other years. HOWEVER, I think we might be able to get the 2001-2005 data from 'PersonsByOccupationalStatusIndustry.csv' - the column names are basically identical to to the column names from the socioeconomy csv, but it might not be comparable to the other years since it's not in the same csv. Because of that, I don't think we should include it - but if we have time, we can.

We went with the unnumbered columns (of the duplicate ones), because they were the only ones containing data for 2009, 2013 and 2017. Both the numbered and unnumbered columns had data for 2021, but we went with the unnumbered ones to match the previous years.

Use socio_2009, socio_2013, socio_2017 and socio_2021 to outerjoin with the final year-dataframes.

Note: Is total calculated correctly? Or should it be summed across all columns, incl. the ones we are not using? Edit: Now it is!

## PersonsByCitizenshipSexAge

In [9]:
cb.inspect_citizenship_duplicates(2001, 20)
cb.inspect_citizenship_duplicates(2005, 20)
cb.inspect_citizenship_duplicates(2009, 20)
cb.inspect_citizenship_duplicates(2013, 20)
cb.inspect_citizenship_duplicates(2017, 20)
cb.inspect_citizenship_duplicates(2021, 20)


CITIZENSHIP DUPLICATE INSPECTION - Year 2001


01. []:
  Found 64 columns
  Citizenship variants with duplicates: 1
  Citizenship variants with single column: 0

  Analyzing 1 duplicate pairs:

  Citizenship: 01.Danmark
    Version 1 (with space):
      Non-null: 0/58, Non-zero: 58, Sum: 0.00, Mean: nan
      Sample values: [nan, nan, nan]
    Version 2 (no space):
      Non-null: 58/58, Non-zero: 58, Sum: 14818.98, Mean: 255.50
      Sample values: [364.91, 188.0, 449.14]
    Version 3 (with space):
      Non-null: 0/58, Non-zero: 58, Sum: 0.00, Mean: nan
      Sample values: [nan, nan, nan]
    Version 4 (no space):
      Non-null: 58/58, Non-zero: 58, Sum: 7784.01, Mean: 134.21
      Sample values: [171.23, 85.0, 145.54]
    Version 5 (with space):
      Non-null: 0/58, Non-zero: 58, Sum: 0.00, Mean: nan
      Sample values: [nan, nan, nan]
    Version 6 (no space):
      Non-null: 58/58, Non-zero: 58, Sum: 4022.02, Mean: 69.35
      Sample values: [68.77, 55.0, 85.77]
    Version 

{'01.': [],
 '02.': [],
 '03.': [],
 '04.': [],
 '05.': [],
 '06.': [],
 '07.': [],
 '08.': [],
 '09.': [],
 '10.': [],
 '11.': [],
 '12.': []}

In [10]:
cb.inspect_citizenship_duplicates_simple(2001, 20)
# Conclusion: Use columns WITHOUT space for 2001


CITIZENSHIP DUPLICATE INSPECTION - Year 2001

Total columns found: 768
  With space (e.g. '01. Danmark'): 384
  Without space (e.g. '01.Danmark'): 384

SUMMARY:
  Columns WITH space:
    Columns with data: 0/384
    Total sum across all: 0.00
  Columns WITHOUT space:
    Columns with data: 384/384
    Total sum across all: 586026.28

CONCLUSION:
  USE columns WITHOUT space (no space after dot)
  IGNORE columns WITH space (they are empty)


In [11]:
cb.inspect_citizenship_duplicates_simple(2005, 20)
# Conclusion: Use columns WITHOUT space for 2005


CITIZENSHIP DUPLICATE INSPECTION - Year 2005

Total columns found: 768
  With space (e.g. '01. Danmark'): 384
  Without space (e.g. '01.Danmark'): 384

SUMMARY:
  Columns WITH space:
    Columns with data: 0/384
    Total sum across all: 0.00
  Columns WITHOUT space:
    Columns with data: 384/384
    Total sum across all: 576873.78

CONCLUSION:
  USE columns WITHOUT space (no space after dot)
  IGNORE columns WITH space (they are empty)


In [12]:
cb.inspect_citizenship_duplicates_simple(2009, 20)
# Conclusion: Use columns WITH space for 2009


CITIZENSHIP DUPLICATE INSPECTION - Year 2009

Total columns found: 768
  With space (e.g. '01. Danmark'): 384
  Without space (e.g. '01.Danmark'): 384

SUMMARY:
  Columns WITH space:
    Columns with data: 384/384
    Total sum across all: 624926.23
  Columns WITHOUT space:
    Columns with data: 0/384
    Total sum across all: 0.00

CONCLUSION:
  USE columns WITH space
  IGNORE columns WITHOUT space (they are empty)


In [13]:
cb.inspect_citizenship_duplicates_simple(2013, 20)
# Conclusion: Use columns WITH space for 2013


CITIZENSHIP DUPLICATE INSPECTION - Year 2013

Total columns found: 768
  With space (e.g. '01. Danmark'): 384
  Without space (e.g. '01.Danmark'): 384

SUMMARY:
  Columns WITH space:
    Columns with data: 384/384
    Total sum across all: 683375.58
  Columns WITHOUT space:
    Columns with data: 0/384
    Total sum across all: 0.00

CONCLUSION:
  USE columns WITH space
  IGNORE columns WITHOUT space (they are empty)


In [14]:
cb.inspect_citizenship_duplicates_simple(2017, 20)
# Conclusion: Use columns WITH space for 2017


CITIZENSHIP DUPLICATE INSPECTION - Year 2017

Total columns found: 768
  With space (e.g. '01. Danmark'): 384
  Without space (e.g. '01.Danmark'): 384

SUMMARY:
  Columns WITH space:
    Columns with data: 384/384
    Total sum across all: 683376.18
  Columns WITHOUT space:
    Columns with data: 0/384
    Total sum across all: 0.00

CONCLUSION:
  USE columns WITH space
  IGNORE columns WITHOUT space (they are empty)


In [15]:
cb.inspect_citizenship_duplicates_simple(2021, 20)
# Conclusion: Use columns WITH space for 2021


CITIZENSHIP DUPLICATE INSPECTION - Year 2021

Total columns found: 768
  With space (e.g. '01. Danmark'): 384
  Without space (e.g. '01.Danmark'): 384

SUMMARY:
  Columns WITH space:
    Columns with data: 384/384
    Total sum across all: 736645.00
  Columns WITHOUT space:
    Columns with data: 0/384
    Total sum across all: 0.00

CONCLUSION:
  USE columns WITH space
  IGNORE columns WITHOUT space (they are empty)


In [16]:
citizenship_2001 = cb.preprocess_citizenship_sex_age(2001)
citizenship_2005 = cb.preprocess_citizenship_sex_age(2005)
citizenship_2009 = cb.preprocess_citizenship_sex_age(2009)
citizenship_2013 = cb.preprocess_citizenship_sex_age(2013)
citizenship_2017 = cb.preprocess_citizenship_sex_age(2017)
citizenship_2021 = cb.preprocess_citizenship_sex_age(2021)

    Citizenship data: Using no-space format for year 2001
    Citizenship data: Using no-space format for year 2005
    Citizenship data: Using with-space format for year 2009
    Citizenship data: Using with-space format for year 2013
    Citizenship data: Using with-space format for year 2017
    Citizenship data: Using with-space format for year 2021


In [17]:
print(f"{'='*80}\n  2001\n")
print(citizenship_2001.info())
print(f"\n{'='*80}\n  2005\n")
print(citizenship_2005.info())
print(f"\n{'='*80}\n  2009\n")
print(citizenship_2009.info())
print(f"\n{'='*80}\n  2013\n")
print(citizenship_2013.info())
print(f"\n{'='*80}\n  2017\n")
print(citizenship_2017.info())
print(f"\n{'='*80}\n  2021\n")
print(citizenship_2021.info())

  2001

<class 'pandas.core.frame.DataFrame'>
Index: 58 entries, 0 to 57
Data columns (total 36 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   Gruppe                                 58 non-null     object 
 1   ValgstedId                             58 non-null     object 
 2   KredsNr                                58 non-null     object 
 3   StorKredsNr                            58 non-null     object 
 4   LandsdelsNr                            58 non-null     object 
 5   Sex_Male                               58 non-null     float64
 6   Sex_Female                             58 non-null     float64
 7   Age_0-4 years                          58 non-null     float64
 8   Age_5-9 years                          58 non-null     float64
 9   Age_10-14 years                        58 non-null     float64
 10  Age_15-17 years                        58 non-null     float64
 11  Age_1

### Conclusion

For 2001 and 2005: Use columns without space
For all subsequent years: Use columns with space

Since the socioeconomy data also is different for 2001 and 2005 compared to subsequent years, maybe we should drop those two early years? It also makes sense, since there was a big policy in Denmark that changed the numbers and borders of municipalities in 2007 ('Kommunalreformen'). That could explain why data from before that might be difficult to compare with data after that.

My opinion is that we should probably just drop 2001 and 2005 for the whole project and only focus on 2009, 2013, 2017 and 2021 (and 2025, of course).

Another small note:
While writing the function `preprocess_citizenship_sex_age(year)`, I realised that columns without citizenship in the column name, e.g. 'KV2021 - Antal personer opgjort efter statsborgerskab køn og aldersgrupper_Kvinder 65-69 år' or 'KV2001 - Antal personer opgjort efter statsborgerskab køn og aldersgrupper_Mænd 0-4 år' were empty, so data for total female, total male, and total within age groups is found by summing up data from each citizenship column.

Also, for the year 20009, the column 'Citizenship_Uoplyst' only has NaN values, so this column should be dropped.

## HighestCompletedEducationByAge

In [18]:
cb.inspect_education_duplicates(2001, 20) # Use '70 år -' format
cb.inspect_education_duplicates(2005, 20) # Use '70 år -' format
cb.inspect_education_duplicates(2009, 20) # Use '70 år -' format
cb.inspect_education_duplicates(2013, 20) # Use '70- år' format
cb.inspect_education_duplicates(2017, 20) # Use '70- år' format
cb.inspect_education_duplicates(2021, 20) # Use '70- år' format


EDUCATION DUPLICATE INSPECTION - Year 2001

Age variant '70- år': 8 columns found
  Columns with data: 0/8
  Total sum: 0.00
Age variant '70 år -': 8 columns found
  Columns with data: 8/8
  Total sum: 65076.05

CONCLUSION:
  ✓ Use '70 år -' format (has data)
  ✗ Ignore '70- år' format (empty)

EDUCATION DUPLICATE INSPECTION - Year 2005

Age variant '70- år': 8 columns found
  Columns with data: 0/8
  Total sum: 0.00
Age variant '70 år -': 8 columns found
  Columns with data: 8/8
  Total sum: 51866.98

CONCLUSION:
  ✓ Use '70 år -' format (has data)
  ✗ Ignore '70- år' format (empty)

EDUCATION DUPLICATE INSPECTION - Year 2009

Age variant '70- år': 8 columns found
  Columns with data: 0/8
  Total sum: 0.00
Age variant '70 år -': 8 columns found
  Columns with data: 8/8
  Total sum: 47885.05

CONCLUSION:
  ✓ Use '70 år -' format (has data)
  ✗ Ignore '70- år' format (empty)

EDUCATION DUPLICATE INSPECTION - Year 2013

Age variant '70- år': 8 columns found
  Columns with data: 8/8
  To

In [19]:
education_2001 = cb.preprocess_education(2001)
education_2005 = cb.preprocess_education(2005)
education_2009 = cb.preprocess_education(2009)
education_2013 = cb.preprocess_education(2013)
education_2017 = cb.preprocess_education(2017)
education_2021 = cb.preprocess_education(2021)

    Education data: Using '70 år -' for 70+ age group
    Education data: Using '70 år -' for 70+ age group
    Education data: Using '70 år -' for 70+ age group
    Education data: Using '70- år' for 70+ age group
    Education data: Using '70- år' for 70+ age group
    Education data: Using '70- år' for 70+ age group


In [20]:
print(f"{'='*80}\n  2001\n")
print(education_2001.info())
print(f"\n{'='*80}\n  2005\n")
print(education_2005.info())
print(f"\n{'='*80}\n  2009\n")
print(education_2009.info())
print(f"\n{'='*80}\n  2013\n")
print(education_2013.info())
print(f"\n{'='*80}\n  2017\n")
print(education_2017.info())
print(f"\n{'='*80}\n  2021\n")
print(education_2021.info())

  2001

<class 'pandas.core.frame.DataFrame'>
Index: 58 entries, 0 to 57
Data columns (total 24 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   Gruppe                                              58 non-null     object 
 1   ValgstedId                                          58 non-null     object 
 2   KredsNr                                             58 non-null     object 
 3   StorKredsNr                                         58 non-null     object 
 4   LandsdelsNr                                         58 non-null     object 
 5   Education_18-29 years_Primary and lower secondary   58 non-null     float64
 6   Education_30-70+ years_Primary and lower secondary  58 non-null     float64
 7   Education_18-29 years_General upper secondary       58 non-null     float64
 8   Education_30-70+ years_General upper secondary      58 non-null     float64
 9 

In [21]:
education_2021.head(10)

,Gruppe,ValgstedId,KredsNr,StorKredsNr,LandsdelsNr,Education_18-29 years_Primary and lower secondary,Education_30-70+ years_Primary and lower secondary,Education_18-29 years_General upper secondary,Education_30-70+ years_General upper secondary,Education_18-29 years_Vocational upper secondary,...,Education_30-70+ years_Short higher education,Education_18-29 years_Medium higher education,Education_30-70+ years_Medium higher education,Education_18-29 years_Long higher education,Education_30-70+ years_Long higher education,Education_18-29 years_Not specified,Education_30-70+ years_Not specified,Education_Total 18-29 years,Education_Total 30-70+ years,Education_Total
0,101001,101001,1,1,1,453.0,885.0,1243.0,569.0,153.0,...,401.0,415.0,1927.0,1433.0,3497.0,21.0,154.0,4060.0,8982.0,13042.0
1,101002,101002,1,1,1,249.0,653.0,519.0,323.0,70.0,...,224.0,195.0,1014.0,577.0,1834.0,17.0,106.0,1837.0,5145.0,6982.0
2,101003,101003,1,1,1,447.0,779.0,1169.0,583.0,126.0,...,387.0,362.0,1964.0,1205.0,3930.0,16.0,139.0,3615.0,9204.0,12819.0
3,101005,101005,1,1,1,438.0,1423.0,1138.0,553.0,167.0,...,403.0,718.0,1709.0,1266.0,2814.0,21.0,220.0,4251.0,9197.0,13448.0
4,101006,101006,1,1,1,479.0,1087.0,980.0,431.0,177.0,...,374.0,407.0,1450.0,1115.0,2466.0,17.0,169.0,3532.0,7720.0,11252.0
5,101007,101007,2,1,1,638.0,2013.0,896.0,476.0,148.0,...,424.0,330.0,1241.0,555.0,1710.0,15.0,239.0,3010.0,8578.0,11588.0
6,101008,101010;101008,2,1,1,991.0,2873.0,2264.0,1117.0,404.0,...,1025.0,915.0,3479.0,2453.0,6417.0,42.0,433.0,8181.0,19681.0,27862.0
7,101009,101057;101009,2,1,1,1166.0,2078.0,2937.0,844.0,592.0,...,874.0,1224.0,2656.0,2232.0,4833.0,62.0,349.0,9320.0,15055.0,24375.0
8,101011,101011,3,1,1,923.0,1575.0,1775.0,889.0,220.0,...,451.0,700.0,1977.0,2012.0,4642.0,73.0,323.0,6424.0,11563.0,17987.0
9,101012,101012,3,1,1,234.0,427.0,605.0,394.0,60.0,...,212.0,185.0,1148.0,553.0,2655.0,13.0,99.0,1776.0,5719.0,7495.0


### Conclusion

Again, there are similarly named columns - but this time only for the last age group 70+, so we just check those. And once again, only one of the two actually contains any data - but it differs from year to year.

For 2001, 2005 and 2009: Use '70 år -' format.

For 2013, 2017 and 2021: Use '70- år' format

We can consider only using the two extremes when we combine the final df - 'Primary and lower secondary' and 'Long higher education' (both of them for both under and over 30)


## ImmigrantsDescendantsByCountryOfOrigin


In [22]:
cb.inspect_immigrants_duplicates(2001, 20) # use columns without space
cb.inspect_immigrants_duplicates(2005, 20) # use columns without space
cb.inspect_immigrants_duplicates(2009, 20) # use columns with space
cb.inspect_immigrants_duplicates(2013, 20) # use columns with space
cb.inspect_immigrants_duplicates(2017, 20) # use columns with space
cb.inspect_immigrants_duplicates(2021, 20) # use columns with space


IMMIGRANTS/DESCENDANTS DUPLICATE INSPECTION - Year 2001

Total columns found: 704
  With space (e.g. '01. Danmark'): 352
  Without space (e.g. '01.Danmark'): 352

SUMMARY:
  Columns WITH space:
    Columns with data: 0/352
    Total sum across all: 0.00
  Columns WITHOUT space:
    Columns with data: 352/352
    Total sum across all: 95053.72

CONCLUSION:
  USE columns WITHOUT space (no space after dot)
  IGNORE columns WITH space (they are empty)

IMMIGRANTS/DESCENDANTS DUPLICATE INSPECTION - Year 2005

Total columns found: 704
  With space (e.g. '01. Danmark'): 352
  Without space (e.g. '01.Danmark'): 352

SUMMARY:
  Columns WITH space:
    Columns with data: 0/352
    Total sum across all: 0.00
  Columns WITHOUT space:
    Columns with data: 352/352
    Total sum across all: 105601.61

CONCLUSION:
  USE columns WITHOUT space (no space after dot)
  IGNORE columns WITH space (they are empty)

IMMIGRANTS/DESCENDANTS DUPLICATE INSPECTION - Year 2009

Total columns found: 704
  With spa

In [23]:
immigrants_2001 = cb.preprocess_immigrants(2001)
immigrants_2005 = cb.preprocess_immigrants(2005)
immigrants_2009 = cb.preprocess_immigrants(2009)
immigrants_2013 = cb.preprocess_immigrants(2013)
immigrants_2017 = cb.preprocess_immigrants(2017)
immigrants_2021 = cb.preprocess_immigrants(2021)

    Immigrants/descendants data: Using no-space format for year 2001
    Immigrants/descendants data: Using no-space format for year 2005
    Immigrants/descendants data: Using with-space format for year 2009
    Immigrants/descendants data: Using with-space format for year 2013
    Immigrants/descendants data: Using with-space format for year 2017
    Immigrants/descendants data: Using with-space format for year 2021


In [24]:
print(f"{'='*80}\n  2001\n")
print(immigrants_2001.info())
print(f"\n{'='*80}\n  2005\n")
print(immigrants_2005.info())
print(f"\n{'='*80}\n  2009\n")
print(immigrants_2009.info())
print(f"\n{'='*80}\n  2013\n")
print(immigrants_2013.info())
print(f"\n{'='*80}\n  2017\n")
print(immigrants_2017.info())
print(f"\n{'='*80}\n  2021\n")
print(immigrants_2021.info())

  2001

<class 'pandas.core.frame.DataFrame'>
Index: 58 entries, 0 to 57
Data columns (total 18 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Gruppe                                58 non-null     object 
 1   ValgstedId                            58 non-null     object 
 2   KredsNr                               58 non-null     object 
 3   StorKredsNr                           58 non-null     object 
 4   LandsdelsNr                           58 non-null     object 
 5   Immigrants_Denmark                    0 non-null      float64
 6   Immigrants_Nordic                     58 non-null     float64
 7   Immigrants_Turkey                     58 non-null     float64
 8   Immigrants_Former Yugoslavia          58 non-null     float64
 9   Immigrants_Old EU countries           58 non-null     float64
 10  Immigrants_New EU countries           58 non-null     float64
 11  Immigrants_Other E

In [25]:
immigrants_2021.head(10)

,Gruppe,ValgstedId,KredsNr,StorKredsNr,LandsdelsNr,Immigrants_Denmark,Immigrants_Nordic,Immigrants_Turkey,Immigrants_Former Yugoslavia,Immigrants_Old EU countries,Immigrants_New EU countries,Immigrants_Other Europe,Immigrants_Africa,Immigrants_North America,Immigrants_South and Central America,Immigrants_Asia and Oceania,Immigrants_Not specified,Immigrants_Total
0,101001,101001,1,1,1,NaN,388.0,40.0,80.0,895.0,275.0,88.0,201.0,168.0,161.0,586.0,1.0,2883.0
1,101002,101002,1,1,1,NaN,148.0,57.0,78.0,360.0,153.0,77.0,120.0,58.0,88.0,525.0,0.0,1664.0
2,101003,101003,1,1,1,NaN,359.0,52.0,39.0,729.0,189.0,83.0,128.0,151.0,212.0,560.0,1.0,2503.0
3,101005,101005,1,1,1,NaN,303.0,136.0,306.0,799.0,406.0,166.0,307.0,194.0,214.0,1113.0,3.0,3947.0
4,101006,101006,1,1,1,NaN,236.0,94.0,199.0,579.0,326.0,118.0,301.0,76.0,130.0,984.0,0.0,3043.0
5,101007,101007,2,1,1,NaN,204.0,320.0,170.0,414.0,442.0,51.0,533.0,55.0,125.0,1537.0,11.0,3862.0
6,101008,101008;101010,2,1,1,NaN,746.0,492.0,327.0,1269.0,663.0,224.0,783.0,219.0,322.0,2396.0,8.0,7449.0
7,101009,101009;101057,2,1,1,NaN,785.0,316.0,398.0,1660.0,1796.0,392.0,676.0,225.0,500.0,3454.0,2.0,10204.0
8,101011,101011,3,1,1,NaN,903.0,130.0,148.0,1575.0,502.0,164.0,528.0,327.0,353.0,1216.0,4.0,5850.0
9,101012,101012,3,1,1,NaN,229.0,24.0,31.0,574.0,124.0,57.0,63.0,136.0,77.0,274.0,0.0,1589.0


### Conclusion
Once again, we have different formats for [2001, 2005] and all other years.

And for obvious reasons, the Denmark column is all-null for all years (no immigrants have immigrated to Denmark from Denmark...), so these columns should be dropped.

## HousingTenure

We might only want to get amount of and no. of people living in either'01. Ejerbolig': 'Owner-occupied housing (ejer)', '02. Andelsbolig': 'Cooperative housing (andel)', '03. Almennyttig bolig': 'Social housing (almen)' or '05. Privat lejerbolig': 'Private rental' and ignore the rest

In [26]:
cb.inspect_housingtenure_duplicates(2001, 20) # use columns with space
cb.inspect_housingtenure_duplicates(2005, 20) # use columns with space
cb.inspect_housingtenure_duplicates(2009, 20) # use columns with space
cb.inspect_housingtenure_duplicates(2013, 20) # use columns with underscore
cb.inspect_housingtenure_duplicates(2017, 20) # use columns with underscore
cb.inspect_housingtenure_duplicates(2021, 20) # use columns with underscore


HOUSING TENURE DUPLICATE INSPECTION - Year 2001

Total columns found: 26

  Spaced columns     (e.g. 'Antal boliger'):   14
  Underscore columns (e.g. 'Antal_boliger'):  12

SUMMARY:
  Columns using SPACE ('Antal boliger' or 'Antal personer'):
    Columns with data: 14/14
    Total sum across all: 910762.02
  Columns using UNDERSCORE ('Antal_boliger' or 'Antal_personer'):
    Columns with data: 0/12
    Total sum across all: 0.00

CONCLUSION:
  USE columns with SPACES ('Antal boliger', 'Antal personer').
  IGNORE columns with underscores (they appear empty).



HOUSING TENURE DUPLICATE INSPECTION - Year 2005

Total columns found: 26

  Spaced columns     (e.g. 'Antal boliger'):   14
  Underscore columns (e.g. 'Antal_boliger'):  12

SUMMARY:
  Columns using SPACE ('Antal boliger' or 'Antal personer'):
    Columns with data: 14/14
    Total sum across all: 889999.96
  Columns using UNDERSCORE ('Antal_boliger' or 'Antal_personer'):
    Columns with data: 0/12
    Total sum across all: 0.

In [27]:
tenure_2001 = cb.preprocess_housingtenure(2001) # use columns with space
tenure_2005 = cb.preprocess_housingtenure(2005) # use columns with space
tenure_2009 = cb.preprocess_housingtenure(2009) # use columns with space
tenure_2013 = cb.preprocess_housingtenure(2013) # use columns with underscore
tenure_2017 = cb.preprocess_housingtenure(2017) # use columns with underscore
tenure_2021 = cb.preprocess_housingtenure(2021) # use columns with underscore

    Housing tenure data: Using spaced format for year 2001
    Housing tenure data: Using spaced format for year 2005
    Housing tenure data: Using spaced format for year 2009
    Housing tenure data: Using underscored format for year 2013
    Housing tenure data: Using underscored format for year 2017
    Housing tenure data: Using underscored format for year 2021


In [28]:
print(f"{'='*80}\n  2001\n")
print(tenure_2001.info())
print(f"\n{'='*80}\n  2005\n")
print(tenure_2005.info())
print(f"\n{'='*80}\n  2009\n")
print(tenure_2009.info())
print(f"\n{'='*80}\n  2013\n")
print(tenure_2013.info())
print(f"\n{'='*80}\n  2017\n")
print(tenure_2017.info())
print(f"\n{'='*80}\n  2021\n")
print(tenure_2021.info())

  2001

<class 'pandas.core.frame.DataFrame'>
Index: 58 entries, 0 to 57
Data columns (total 15 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Gruppe                                58 non-null     object 
 1   ValgstedId                            58 non-null     object 
 2   KredsNr                               58 non-null     object 
 3   StorKredsNr                           58 non-null     object 
 4   LandsdelsNr                           58 non-null     object 
 5   Housing tenure_Homes_Ejer             58 non-null     float64
 6   Housing tenure_Homes_Andel            58 non-null     float64
 7   Housing tenure_Homes_Almen            58 non-null     float64
 8   Housing tenure_Homes_Private rental   58 non-null     float64
 9   Housing tenure_People_Ejer            58 non-null     float64
 10  Housing tenure_People_Andel           58 non-null     float64
 11  Housing tenure_Peo

In [29]:
tenure_2021.head(10)

,Gruppe,ValgstedId,KredsNr,StorKredsNr,LandsdelsNr,Housing tenure_Homes_Ejer,Housing tenure_Homes_Andel,Housing tenure_Homes_Almen,Housing tenure_Homes_Private rental,Housing tenure_People_Ejer,Housing tenure_People_Andel,Housing tenure_People_Almen,Housing tenure_People_Private rental,Housing tenure_Total people,Housing tenure_Total homes
0,101001,101001,1,1,1,1248.0,2823.0,1026.0,861.0,2719.0,5490.0,1703.0,1526.0,11438.0,5958.0
1,101002,101002,1,1,1,1097.0,1550.0,562.0,373.0,2503.0,2933.0,918.0,663.0,7017.0,3582.0
2,101003,101003,1,1,1,1181.0,3492.0,254.0,676.0,2834.0,7613.0,424.0,1322.0,12193.0,5603.0
3,101005,101005,1,1,1,1137.0,3335.0,1196.0,576.0,2550.0,6783.0,1945.0,1072.0,12350.0,6244.0
4,101006,101006,1,1,1,1276.0,2378.0,1354.0,591.0,2845.0,4483.0,2578.0,1117.0,11023.0,5599.0
5,101007,101007,2,1,1,1511.0,1769.0,2721.0,402.0,4063.0,3167.0,4830.0,840.0,12900.0,6403.0
6,101008,101010;101008,2,1,1,2780.0,6860.0,2828.0,1484.0,6270.0,13391.0,5514.0,2847.0,28022.0,13952.0
7,101009,101057;101009,2,1,1,4357.0,432.0,2897.0,552.0,11579.0,829.0,6107.0,1298.0,19813.0,8238.0
8,101011,101011,3,1,1,1517.0,2162.0,112.0,1216.0,3073.0,4524.0,185.0,2385.0,10167.0,5007.0
9,101012,101012,3,1,1,1155.0,976.0,0.0,706.0,2928.0,2002.0,0.0,1514.0,6444.0,2837.0


## HousingType

DST says:

1. Enfamilie-, stue-, parcelhus (BBR-koder (110 Stuehus til landbrugsejendom og 120 Fritliggende enfamilieshus (parcelhus)).
2. Række-, kæde- dobbelthus (BBR-kode 130 Række-, kæde- eller dobbelthus (lodret adskillelse mellem enhederne)).
3. Etagebolig (BBR-kode 140 Etageboligbebyggelse (flerfamiliehus, herunder tofamiliehus (vandret adskillelse mellem enhederne))).
4. Kollegiebolig (BBR-kode 150 Kollegium)
6. Fritidshus (BBR-koder 510 Sommerhus. 540 Kolonihavehus.590 Anden enhed til fritidsformål)

In our data, we have:
- 01. Stuehuse og parcelhuse (detached houses/villas)
- 02. Række- og kædehuse (terraced houses)
- 03. Flerfamiliehuse (multi-family/block of flats)
- 04. Kollegier (student halls)
- 05. Sommerhuse (holiday homes)

So 'flerfamiliehuse' is apartment buildings/multi-storey residential buildings - or just two-family-houses, where the two homes are separated horizontally (as opposed to vertically, which would go under 'række- og kædehuse')
We will assume that 05. sommerhuse also covers kolonihavehuse, which we have a few of in cph.

I think we should put detached and terraced houses together in one column, blocks in one column. And ignore the rest.

In [30]:
cb.inspect_housingtype_duplicates(2001, 20) # use columns with space
cb.inspect_housingtype_duplicates(2005, 20) # use columns with space
cb.inspect_housingtype_duplicates(2009, 20) # use columns with space
cb.inspect_housingtype_duplicates(2013, 20) # use columns with underscore
cb.inspect_housingtype_duplicates(2017, 20) # use columns with underscore
cb.inspect_housingtype_duplicates(2021, 20) # use columns with underscore


HOUSING TYPE DUPLICATE INSPECTION - Year 2001

Total columns found: 26

  Spaced columns     (e.g. 'Antal boliger'):   14
  Underscore columns (e.g. 'Antal_boliger'):  12

SUMMARY:
  Columns using SPACE ('Antal boliger' or 'Antal personer'):
    Columns with data: 14/14
    Total sum across all: 910761.96
  Columns using UNDERSCORE ('Antal_boliger' or 'Antal_personer'):
    Columns with data: 0/12
    Total sum across all: 0.00

CONCLUSION:
  USE columns with SPACES ('Antal boliger', 'Antal personer').
  IGNORE columns with underscores (they appear empty).



HOUSING TYPE DUPLICATE INSPECTION - Year 2005

Total columns found: 26

  Spaced columns     (e.g. 'Antal boliger'):   14
  Underscore columns (e.g. 'Antal_boliger'):  12

SUMMARY:
  Columns using SPACE ('Antal boliger' or 'Antal personer'):
    Columns with data: 14/14
    Total sum across all: 889999.89
  Columns using UNDERSCORE ('Antal_boliger' or 'Antal_personer'):
    Columns with data: 0/12
    Total sum across all: 0.00



In [31]:
type_2001 = cb.preprocess_housingtype(2001) # use columns with space
type_2005 = cb.preprocess_housingtype(2005) # use columns with space
type_2009 = cb.preprocess_housingtype(2009) # use columns with space
type_2013 = cb.preprocess_housingtype(2013) # use columns with underscore
type_2017 = cb.preprocess_housingtype(2017) # use columns with underscore
type_2021 = cb.preprocess_housingtype(2021) # use columns with underscore

    Housing type data: Using spaced format for year 2001
    Housing type data: Using spaced format for year 2005
    Housing type data: Using spaced format for year 2009
    Housing type data: Using underscored format for year 2013
    Housing type data: Using underscored format for year 2017
    Housing type data: Using underscored format for year 2021


In [32]:
print(f"{'='*80}\n  2001\n")
print(type_2001.info())
print(f"\n{'='*80}\n  2005\n")
print(type_2005.info())
print(f"\n{'='*80}\n  2009\n")
print(type_2009.info())
print(f"\n{'='*80}\n  2013\n")
print(type_2013.info())
print(f"\n{'='*80}\n  2017\n")
print(type_2017.info())
print(f"\n{'='*80}\n  2021\n")
print(type_2021.info())

  2001

<class 'pandas.core.frame.DataFrame'>
Index: 58 entries, 0 to 57
Data columns (total 21 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   Gruppe                                   58 non-null     object 
 1   ValgstedId                               58 non-null     object 
 2   KredsNr                                  58 non-null     object 
 3   StorKredsNr                              58 non-null     object 
 4   LandsdelsNr                              58 non-null     object 
 5   Housing type_Homes_Detached house        58 non-null     float64
 6   Housing type_Homes_Terraced house        58 non-null     float64
 7   Housing type_Homes_Apartment buildings   58 non-null     float64
 8   Housing type_Homes_Student halls         58 non-null     float64
 9   Housing type_Homes_Holiday homes         58 non-null     float64
 10  Housing type_Homes_Other homes           58 non-n

In [33]:
type_2017.head(10)

,Gruppe,ValgstedId,KredsNr,StorKredsNr,LandsdelsNr,Housing type_Homes_Detached house,Housing type_Homes_Terraced house,Housing type_Homes_Apartment buildings,Housing type_Homes_Student halls,Housing type_Homes_Holiday homes,...,Housing type_Homes_No BBR information,Housing type_People_Detached house,Housing type_People_Terraced house,Housing type_People_Apartment buildings,Housing type_People_Student halls,Housing type_People_Holiday homes,Housing type_People_Other homes,Housing type_People_No BBR information,Housing type_Total people,Housing type_Total homes
0,101001,101001,1,1,1,1.49,35.05,7115.07,0.00,0.75,...,NaN,7.46,107.37,13556.81,0.00,1.49,76.06,NaN,13749.19,7188.90
1,101002,101002,1,1,1,218.00,81.00,3748.00,7.00,1.00,...,NaN,812.00,232.00,6782.00,7.00,1.00,121.00,NaN,7955.00,4075.00
2,101003,101003,1,1,1,18.80,5.48,7624.50,206.77,0.78,...,NaN,65.79,12.53,15383.70,217.73,0.78,112.00,NaN,15792.53,7900.97
3,101005,101005,1,1,1,5.00,141.00,7916.00,162.00,4.00,...,NaN,10.00,515.00,14500.00,229.00,4.00,82.00,NaN,15340.00,8284.00
4,101006,101006,1,1,1,14.00,364.00,6486.00,85.00,0.00,...,NaN,40.00,1158.00,11781.00,91.00,0.00,137.00,NaN,13207.00,7013.00
5,101007,101007,2,1,1,1138.98,475.69,4961.62,119.11,1.49,...,NaN,3198.06,1452.38,9216.77,141.44,2.23,64.77,NaN,14075.65,6739.32
6,101008,101010;101008,2,1,1,7.00,135.00,15039.00,565.00,9.00,...,NaN,19.00,468.00,29609.00,690.00,57.00,396.00,NaN,31239.00,15814.00
7,101009,101057;101009,2,1,1,1343.02,547.31,5773.38,1456.89,20.51,...,NaN,3738.94,1586.62,12362.23,1691.56,45.77,217.23,NaN,19642.35,9259.68
8,101011,101011,3,1,1,6.00,8.00,7257.00,311.00,4.00,...,NaN,12.00,20.00,14634.00,442.00,6.00,4147.00,NaN,19261.00,7724.00
9,101012,101012,3,1,1,5.00,430.00,3636.00,1.00,1.00,...,NaN,14.00,1282.00,7641.00,60.00,3.00,60.00,NaN,9060.00,4097.00


In [34]:
type_2021.describe()

,Housing type_Homes_Detached house,Housing type_Homes_Terraced house,Housing type_Homes_Apartment buildings,Housing type_Homes_Student halls,Housing type_Homes_Holiday homes,Housing type_Homes_Other homes,Housing type_Homes_No BBR information,Housing type_People_Detached house,Housing type_People_Terraced house,Housing type_People_Apartment buildings,Housing type_People_Student halls,Housing type_People_Holiday homes,Housing type_People_Other homes,Housing type_People_No BBR information,Housing type_Total people,Housing type_Total homes
count,58.000000,58.000000,58.000000,58.000000,58.000000,58.000000,0.0,58.000000,58.000000,58.000000,58.000000,58.000000,58.000000,0.0,58.000000,58.000000
mean,268.689655,170.086207,5633.810345,163.500000,1.706897,32.103448,NaN,827.189655,502.155172,10991.413793,208.948276,3.189655,167.879310,NaN,12700.775862,6269.896552
std,419.681634,226.612296,2486.069289,331.866717,3.815980,36.721660,NaN,1277.016609,665.350872,4862.007220,414.462087,8.038192,503.385907,NaN,5285.771022,2643.388453
min,0.000000,0.000000,1614.000000,0.000000,0.000000,0.000000,NaN,0.000000,0.000000,2787.000000,0.000000,0.000000,0.000000,NaN,4468.000000,2497.000000
25%,4.000000,12.250000,4015.250000,0.000000,0.000000,9.000000,NaN,13.250000,36.250000,8083.250000,0.000000,0.000000,20.750000,NaN,9372.500000,4702.750000
50%,63.500000,88.500000,5296.500000,18.000000,0.000000,16.000000,NaN,221.000000,266.500000,10633.000000,48.000000,0.000000,53.500000,NaN,11982.500000,5699.000000
75%,365.000000,275.500000,7129.500000,170.750000,2.000000,47.000000,NaN,1062.500000,844.500000,13703.750000,221.250000,2.750000,159.000000,NaN,14489.750000,7392.750000
max,1956.000000,1312.000000,15867.000000,1841.000000,19.000000,148.000000,NaN,5569.000000,4016.000000,31287.000000,2232.000000,47.000000,3831.000000,NaN,33643.000000,16761.000000


In [35]:
type_2021[type_2021['Housing type_Homes_Detached house'] == type_2021['Housing type_Homes_Detached house'].max()]

,Gruppe,ValgstedId,KredsNr,StorKredsNr,LandsdelsNr,Housing type_Homes_Detached house,Housing type_Homes_Terraced house,Housing type_Homes_Apartment buildings,Housing type_Homes_Student halls,Housing type_Homes_Holiday homes,...,Housing type_Homes_No BBR information,Housing type_People_Detached house,Housing type_People_Terraced house,Housing type_People_Apartment buildings,Housing type_People_Student halls,Housing type_People_Holiday homes,Housing type_People_Other homes,Housing type_People_No BBR information,Housing type_Total people,Housing type_Total homes
7,101009,101057;101009,2,1,1,1956.0,1312.0,8413,1841.0,19.0,...,NaN,5569.0,4016.0,19050,2232.0,32.0,163.0,NaN,31062.0,13616.0


### Conclusion

In the final combined df:
- Combine detached and terraced into one column
- Use apartments, the combined column and the total columns, but none of the other

## Income, benefit type, cars, crime

In [36]:
simple_2001 = cb.preprocess_simple(2001) 
simple_2005 = cb.preprocess_simple(2005) 
simple_2009 = cb.preprocess_simple(2009) 
simple_2013 = cb.preprocess_simple(2013)
simple_2017 = cb.preprocess_simple(2017)
simple_2021 = cb.preprocess_simple(2021)

    Extracting from PersonsByBenefitType.csv...
    Extracting from CrimeArrests.csv...
    Extracting from HouseholdIncomes.csv...
    Extracting from HouseholdsCarOwnership.csv...
    Extracting from PersonsByBenefitType.csv...
    Extracting from CrimeArrests.csv...
    Extracting from HouseholdIncomes.csv...
    Extracting from HouseholdsCarOwnership.csv...
    Extracting from PersonsByBenefitType.csv...
    Extracting from CrimeArrests.csv...
    Extracting from HouseholdIncomes.csv...
    Extracting from HouseholdsCarOwnership.csv...
    Extracting from PersonsByBenefitType.csv...
    Extracting from CrimeArrests.csv...
    Extracting from HouseholdIncomes.csv...
    Extracting from HouseholdsCarOwnership.csv...
    Extracting from PersonsByBenefitType.csv...
    Extracting from CrimeArrests.csv...
    Extracting from HouseholdIncomes.csv...
    Extracting from HouseholdsCarOwnership.csv...
    Extracting from PersonsByBenefitType.csv...
    Extracting from CrimeArrests.csv...
  

In [37]:
print(f"{'='*80}\n  2001\n")
print(simple_2001.info())
print(f"\n{'='*80}\n  2005\n")
print(simple_2005.info())
print(f"\n{'='*80}\n  2009\n")
print(simple_2009.info())
print(f"\n{'='*80}\n  2013\n")
print(simple_2013.info())
print(f"\n{'='*80}\n  2017\n")
print(simple_2017.info())
print(f"\n{'='*80}\n  2021\n")
print(simple_2021.info())

  2001

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58 entries, 0 to 57
Data columns (total 17 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Gruppe                              58 non-null     object 
 1   ValgstedId                          58 non-null     object 
 2   KredsNr                             58 non-null     object 
 3   StorKredsNr                         58 non-null     object 
 4   LandsdelsNr                         58 non-null     object 
 5   Benefit type_Kontanthjaelp          0 non-null      float64
 6   Benefit type_Foertidspension        0 non-null      float64
 7   Benefit type_Folkepension           0 non-null      float64
 8   Benefit type_Modtager ikke ydelser  0 non-null      float64
 9   Benefit type_Antal personer i alt   0 non-null      float64
 10  Crime_Number of arrests             0 non-null      float64
 11  Income_Median household income      0 n

In [38]:
simple_2021.head(10)

,Gruppe,ValgstedId,KredsNr,StorKredsNr,LandsdelsNr,Benefit type_Kontanthjaelp,Benefit type_Foertidspension,Benefit type_Folkepension,Benefit type_Modtager ikke ydelser,Benefit type_Antal personer i alt,Crime_Number of arrests,Income_Median household income,Income_Total households,Income_Gross income top20pct,Car_Households with no car,Car_Households with 1 car,Car_Households with 2+ cars
0,101001,101001,1,1,1,259,244,1646,11236,15562,40.0,491202,7928,2353964408,5056,956,2032
1,101002,101002,1,1,1,184,160,1307,5669,8378,29.0,469505,4351,1813280147,2385,658,1256
2,101003,101003,1,1,1,170,177,1579,11748,15846,41.0,542786,7547,2551615423,4591,807,2229
3,101005,101005,1,1,1,314,376,2038,10767,15854,68.0,438719,8208,1974027178,5361,1057,1958
4,101006,101006,1,1,1,307,372,1617,9086,13487,38.0,449840,7164,1844543365,4349,954,1802
5,101007,101007,2,1,1,572,510,1997,9065,14265,126.0,401150,7081,1750447556,4173,856,2084
6,101008,101010;101008,2,1,1,892,807,3109,23509,33643,182.0,1070952,16801,4777664075,10037,2049,4675
7,101009,101057;101009,2,1,1,480,551,1795,23792,31062,92.0,997902,14163,3800875512,8164,1193,4259
8,101011,101011,3,1,1,1006,242,1698,15027,20283,639.0,545166,8024,3085311495,5368,916,1838
9,101012,101012,3,1,1,74,65,1271,6748,9176,15.0,649124,4072,2114427532,2045,518,1565


In [39]:
simple_2021.describe()

,Benefit type_Kontanthjaelp,Benefit type_Foertidspension,Benefit type_Folkepension,Benefit type_Modtager ikke ydelser,Benefit type_Antal personer i alt,Crime_Number of arrests,Income_Median household income,Income_Total households,Income_Gross income top20pct,Car_Households with no car,Car_Households with 1 car,Car_Households with 2+ cars
count,58.000000,58.000000,58.000000,58.000000,58.000000,58.000000,5.800000e+01,58.000000,5.800000e+01,58.000000,58.000000,58.000000
mean,335.741379,319.827586,1326.189655,8787.931034,12700.775862,73.396552,5.016165e+05,6271.672414,1.738676e+09,3802.155172,753.689655,1714.034483
std,200.674263,169.927140,688.845779,3912.312727,5285.771022,85.355591,1.490076e+05,2656.540907,8.520362e+08,1749.392324,333.382990,776.124718
min,74.000000,65.000000,383.000000,2969.000000,4468.000000,0.000000,2.764210e+05,2484.000000,4.069871e+08,1090.000000,236.000000,289.000000
25%,181.000000,201.500000,765.000000,6722.500000,9372.500000,34.000000,4.176092e+05,4544.250000,1.203761e+09,2658.000000,518.250000,1164.750000
50%,301.000000,257.000000,1232.000000,8250.500000,11982.500000,56.500000,4.687070e+05,5673.500000,1.521289e+09,3399.500000,726.000000,1598.500000
75%,417.500000,392.000000,1692.750000,9902.000000,14489.750000,91.250000,5.385520e+05,7493.500000,2.201399e+09,4755.000000,909.250000,2106.250000
max,1006.000000,807.000000,3608.000000,23792.000000,33643.000000,639.000000,1.070952e+06,16801.000000,4.777664e+09,10037.000000,2049.000000,4675.000000


### Conclusion
Again, no data for any of the features for 2001 and 2005.

Crime data missing for 2009.

## HousingByYearOfConstruction

In [40]:
cb.inspect_constructionyear_duplicates(2001, 20) # use columns with space, but drop columns with 2000- 
cb.inspect_constructionyear_duplicates(2005, 20) # use columns with space
cb.inspect_constructionyear_duplicates(2009, 20) # use columns with space
cb.inspect_constructionyear_duplicates(2013, 20) # use underscored columns (make sure to get Antal_boliger_Før 1940 and Antal_personer_Før 1940)
cb.inspect_constructionyear_duplicates(2017, 20) # use underscored columns (make sure to get Antal_boliger_Før 1940 and Antal_personer_Før 1940)
cb.inspect_constructionyear_duplicates(2021, 20) # use underscored columns (make sure to get Antal_boliger_Før 1940 and Antal_personer_Før 1940)


CONSTRUCTION YEAR INSPECTION - Year 2001

Total columns found: 30

Columns with SPACE (e.g., 'Antal boliger_...'):
  Total: 16
  With data: 12
    ✓ Antal boliger_1940-1959                                      sum=45285
    ✓ Antal boliger_1960-1969                                      sum=18488
    ✓ Antal boliger_1970-1979                                      sum=20009
    ✓ Antal boliger_1980-1989                                      sum=11522
    ✓ Antal boliger_1990-1999                                      sum=7816
    ✗ Antal boliger_2000-                                          sum=0
    ✓ Antal boliger_Før 1940                                       sum=221616
    ✓ Antal personer_1940-1959                                     sum=74006
    ✓ Antal personer_1960-1969                                     sum=31599
    ✓ Antal personer_1970-1979                                     sum=35223
    ✓ Antal personer_1980-1989                                     sum=29493
    ✓ Antal p

In [41]:
builtyear_2001 = cb.preprocess_builtyear(2001) 
builtyear_2005 = cb.preprocess_builtyear(2005) 
builtyear_2009 = cb.preprocess_builtyear(2009) 
builtyear_2013 = cb.preprocess_builtyear(2013)
builtyear_2017 = cb.preprocess_builtyear(2017)
builtyear_2021 = cb.preprocess_builtyear(2021)

    Construction year data: Using spaced format for year 2001
    Construction year data: Using spaced format for year 2005
    Construction year data: Using spaced format for year 2009
    Construction year data: Using underscored format for year 2013
    Construction year data: Using underscored format for year 2017
    Construction year data: Using underscored format for year 2021


In [42]:
print(f"{'='*80}\n  2001\n")
print(builtyear_2001.info())
print(f"\n{'='*80}\n  2005\n")
print(builtyear_2005.info())
print(f"\n{'='*80}\n  2009\n")
print(builtyear_2009.info())
print(f"\n{'='*80}\n  2013\n")
print(builtyear_2013.info())
print(f"\n{'='*80}\n  2017\n")
print(builtyear_2017.info())
print(f"\n{'='*80}\n  2021\n")
print(builtyear_2021.info())

  2001

<class 'pandas.core.frame.DataFrame'>
Index: 58 entries, 0 to 57
Data columns (total 25 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Gruppe                       58 non-null     object 
 1   ValgstedId                   58 non-null     object 
 2   KredsNr                      58 non-null     object 
 3   StorKredsNr                  58 non-null     object 
 4   LandsdelsNr                  58 non-null     object 
 5   Year built_Homes_Før 1940    58 non-null     float64
 6   Year built_Homes_1940-1959   58 non-null     float64
 7   Year built_Homes_1960-1969   58 non-null     float64
 8   Year built_Homes_1970-1979   58 non-null     float64
 9   Year built_Homes_1980-1989   58 non-null     float64
 10  Year built_Homes_1990-1999   58 non-null     float64
 11  Year built_Homes_2000-2009   0 non-null      float64
 12  Year built_Homes_2010-9999   0 non-null      float64
 13  Year built_Homes_20

In [43]:
builtyear_2021.info()

<class 'pandas.core.frame.DataFrame'>
Index: 58 entries, 0 to 57
Data columns (total 25 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Gruppe                       58 non-null     object 
 1   ValgstedId                   58 non-null     object 
 2   KredsNr                      58 non-null     object 
 3   StorKredsNr                  58 non-null     object 
 4   LandsdelsNr                  58 non-null     object 
 5   Year built_Homes_Før 1940    58 non-null     int64  
 6   Year built_Homes_1940-1959   58 non-null     float64
 7   Year built_Homes_1960-1969   58 non-null     float64
 8   Year built_Homes_1970-1979   58 non-null     float64
 9   Year built_Homes_1980-1989   58 non-null     float64
 10  Year built_Homes_1990-1999   58 non-null     float64
 11  Year built_Homes_2000-2009   58 non-null     float64
 12  Year built_Homes_2010-9999   58 non-null     float64
 13  Year built_Homes_2000-     

We want to split the columns into either two bins, three or four bins. For example 'before 1960', '1960-2010' and '2010-'

In [44]:
old_homes_cols = [
    'Year built_Homes_Før 1940',
    'Year built_Homes_1940-1959'
]
middle_homes_cols = [
    'Year built_Homes_1960-1969',
    'Year built_Homes_1970-1979',
    'Year built_Homes_1980-1989',    
    'Year built_Homes_1990-1999',
    'Year built_Homes_2000-2009',
]
old_people_cols = [
    'Year built_People_Før 1940',
    'Year built_People_1940-1959'
]
middle_people_cols = [
    'Year built_People_1960-1969',
    'Year built_People_1970-1979',
    'Year built_People_1980-1989',    
    'Year built_People_1990-1999',
    'Year built_People_2000-2009',
]

builtyear_2021["Year built_People_Before 1960"] = builtyear_2021[old_people_cols].sum(axis=1)
builtyear_2021["Year built_People_1960-2009"] = builtyear_2021[middle_people_cols].sum(axis=1)
builtyear_2021["Year built_Homes_Before 1960"] = builtyear_2021[old_homes_cols].sum(axis=1)
builtyear_2021["Year built_Homes_1960-2009"] = builtyear_2021[middle_homes_cols].sum(axis=1)

columns_to_keep = [
    "Gruppe",
    "ValgstedId",
    "KredsNr",
    "StorKredsNr",
    "LandsdelsNr",
    "Year built_People_Before 1960",
    "Year built_People_1960-2009",
    "Year built_Homes_Before 1960",
    "Year built_Homes_1960-2009",
    "Year built_People_2010-9999",
    "Year built_Homes_2010-9999",
    "Year built_Total people",
    "Year built_Total homes",
]

builtyear_2021 = builtyear_2021[columns_to_keep]

In [45]:
builtyear_2021.head(10)

,Gruppe,ValgstedId,KredsNr,StorKredsNr,LandsdelsNr,Year built_People_Before 1960,Year built_People_1960-2009,Year built_Homes_Before 1960,Year built_Homes_1960-2009,Year built_People_2010-9999,Year built_Homes_2010-9999,Year built_Total people,Year built_Total homes
0,101001,101001,1,1,1,13417.0,2055.0,6887.0,1117.0,90.0,40.0,15562.0,8044.0
1,101002,101002,1,1,1,6994.0,1235.0,3418.0,809.0,149.0,72.0,8378.0,4299.0
2,101003,101003,1,1,1,14942.0,904.0,7041.0,586.0,0.0,0.0,15846.0,7627.0
3,101005,101005,1,1,1,11810.0,3873.0,6119.0,2167.0,171.0,90.0,15854.0,8376.0
4,101006,101006,1,1,1,10028.0,3240.0,5283.0,1737.0,219.0,85.0,13487.0,7105.0
5,101007,101007,2,1,1,8275.0,5259.0,3950.0,2852.0,731.0,311.0,14265.0,7113.0
6,101008,101010;101008,2,1,1,19096.0,11105.0,9939.0,5533.0,3442.0,1289.0,33643.0,16761.0
7,101009,101009;101057,2,1,1,4452.0,13646.0,1659.0,6621.0,12964.0,5336.0,31062.0,13616.0
8,101011,101011,3,1,1,19560.0,690.0,7725.0,376.0,33.0,22.0,20283.0,8123.0
9,101012,101012,3,1,1,9040.0,136.0,4033.0,95.0,0.0,0.0,9176.0,4128.0


### Conclusion
Difference in data format: 2001, 2005 and 2009 used spaced, subsequent years use underscored.

It looks like the data for 2013 and 2017 is the same - do they have the same data? Check this. We should check this for other collected features as well.

Maybe we should only keep 2013, 2017 and 2021 - and if 2013 and 2017 are the same, then maybe only 2013 and 2021. Or just 2021? It is hard to compare across time since this column also is about time. Obviously, 2009 will not have any data on housing built after 2010.

# Combine everything

- socioeconomy
- citizenship (and age and sex)
- education
- immigrants
- tenure
- type
- simple (contains income, car, benefit type, arrests)
- builtyear (housing construction year)

In [46]:
names = [
    'socio',
    'citizenship',
    'immigrants',
    'education',
    'tenure',
    'type',
    'simple',
    'builtyear'
]
years = [2009, 2013, 2017, 2021]

combined_dfs = {}
for year in years:
    # start with first dataframe
    df_combined = globals()[f'{names[0]}_{year}'].copy()
    # merge subsequent dataframes
    for name in names[1:]:
        df_name = f'{name}_{year}'
        df_combined = df_combined.merge(
            globals()[df_name], 
            on=COMMON_COLS,
            how='outer',
            suffixes=('', f'_{name}')
        )
    combined_dfs[year] = df_combined

df_2009 = combined_dfs[2009]
df_2013 = combined_dfs[2013]
df_2017 = combined_dfs[2017]
df_2021 = combined_dfs[2021]

In [47]:
print(f"\n{'='*80}\n  2009\n")
print(df_2009.info())
print(f"\n{'='*80}\n  2013\n")
print(df_2013.info())
print(f"\n{'='*80}\n  2017\n")
print(df_2017.info())
print(f"\n{'='*80}\n  2021\n")
print(df_2021.info())


  2009

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63 entries, 0 to 62
Columns: 130 entries, Gruppe to Year built_Total homes
dtypes: float64(125), object(5)
memory usage: 64.1+ KB
None

  2013

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63 entries, 0 to 62
Columns: 130 entries, Gruppe to Year built_Total homes
dtypes: float64(125), object(5)
memory usage: 64.1+ KB
None

  2017

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63 entries, 0 to 62
Columns: 130 entries, Gruppe to Year built_Total homes
dtypes: float64(125), object(5)
memory usage: 64.1+ KB
None

  2021

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63 entries, 0 to 62
Columns: 118 entries, Gruppe to Year built_Total homes
dtypes: float64(113), object(5)
memory usage: 58.2+ KB
None


In [48]:
df_2021

,Gruppe,ValgstedId,KredsNr,StorKredsNr,LandsdelsNr,Socioeconomy_Self-employed,Socioeconomy_Top executives,Socioeconomy_Unemployed,Socioeconomy_Total,Sex_Male,...,Car_Households with 1 car,Car_Households with 2+ cars,Year built_People_Before 1960,Year built_People_1960-2009,Year built_Homes_Before 1960,Year built_Homes_1960-2009,Year built_People_2010-9999,Year built_Homes_2010-9999,Year built_Total people,Year built_Total homes
0,101001,101001,1,1,1,563.0,383.0,305.0,15540.0,7331.0,...,956.0,2032.0,13417.0,2055.0,6887.0,1117.0,90.0,40.0,15562.0,8044.0
1,101002,101002,1,1,1,357.0,221.0,137.0,8366.0,3908.0,...,658.0,1256.0,6994.0,1235.0,3418.0,809.0,149.0,72.0,8378.0,4299.0
2,101003,101003,1,1,1,632.0,401.0,277.0,15838.0,7640.0,...,807.0,2229.0,14942.0,904.0,7041.0,586.0,0.0,0.0,15846.0,7627.0
3,101005,101005,1,1,1,396.0,230.0,312.0,15817.0,7648.0,...,1057.0,1958.0,11810.0,3873.0,6119.0,2167.0,171.0,90.0,15854.0,8376.0
4,101006,101006,1,1,1,410.0,246.0,302.0,13473.0,6527.0,...,954.0,1802.0,10028.0,3240.0,5283.0,1737.0,219.0,85.0,13487.0,7105.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
58,147004,147004,10,1,1,437.0,366.0,274.0,14633.0,6807.0,...,997.0,2273.0,10458.0,3690.0,5139.0,1896.0,428.0,354.0,14576.0,7389.0
59,147005,147005,11,1,1,383.0,287.0,302.0,14555.0,6845.0,...,1038.0,2092.0,10537.0,3303.0,5674.0,1874.0,445.0,301.0,14285.0,7849.0
60,147006,147006,11,1,1,344.0,250.0,202.0,12109.0,5532.0,...,1049.0,1960.0,8707.0,3142.0,4435.0,1999.0,289.0,262.0,12138.0,6696.0
61,147007,147007;147008,11,1,1,NaN,NaN,NaN,NaN,11310.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [49]:
common = df_2021[COMMON_COLS]
common.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63 entries, 0 to 62
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Gruppe       63 non-null     object
 1   ValgstedId   63 non-null     object
 2   KredsNr      63 non-null     object
 3   StorKredsNr  63 non-null     object
 4   LandsdelsNr  63 non-null     object
dtypes: object(5)
memory usage: 2.6+ KB


In [50]:
pd.set_option('display.max_rows', None)
print(df_2021[COMMON_COLS])
pd.reset_option('display.max_rows')

    Gruppe     ValgstedId KredsNr StorKredsNr LandsdelsNr
0   101001         101001       1           1           1
1   101002         101002       1           1           1
2   101003         101003       1           1           1
3   101005         101005       1           1           1
4   101006         101006       1           1           1
5   101007         101007       2           1           1
6   101008  101008;101010       2           1           1
7   101008  101010;101008       2           1           1
8   101009         101009       2           1           1
9   101009  101009;101057       2           1           1
10  101009  101057;101009       2           1           1
11  101011         101011       3           1           1
12  101012         101012       3           1           1
13  101013         101013       3           1           1
14  101014         101014       3           1           1
15  101014  101057;101014     2;3           1           1
16  101016    

There are duplicate 'Gruppe' values - we only want 58 rows, but we get 63, because we're merging on several common columns, and some of them are different. It is only the CSV 'PersonsBySocioeconomyIndustry.csv' that is causing some of them being different, I think. We'll merge just on 'Gruppe' instead, and then deal with duplicate common columns after.

In [51]:
names = [
    'socio',
    'citizenship',
    'immigrants',
    'education',
    'tenure',
    'type',
    'simple',
    'builtyear'
]
years = [2009, 2013, 2017, 2021]

combined_dfs = {}
for year in years:
    # start with first dataframe
    df_combined = globals()[f'{names[0]}_{year}'].copy()
    # merge subsequent dataframes
    for name in names[1:]:
        df_name = f'{name}_{year}'
        df_combined = df_combined.merge(
            globals()[df_name], 
            on='Gruppe',
            how='outer',
            suffixes=('', f'_{name}')
        )
    combined_dfs[year] = df_combined

df_2009 = combined_dfs[2009]
df_2013 = combined_dfs[2013]
df_2017 = combined_dfs[2017]
df_2021 = combined_dfs[2021]

print(f"df_2021 shape: {df_2021.shape}")
print(f"Unique Gruppe values: {df_2021['Gruppe'].nunique()}")

df_2021 shape: (58, 146)
Unique Gruppe values: 58


In [52]:
print(df_2021.columns)

Index(['Gruppe', 'ValgstedId', 'KredsNr', 'StorKredsNr', 'LandsdelsNr',
       'Socioeconomy_Self-employed', 'Socioeconomy_Top executives',
       'Socioeconomy_Unemployed', 'Socioeconomy_Total',
       'ValgstedId_citizenship',
       ...
       'StorKredsNr_builtyear', 'LandsdelsNr_builtyear',
       'Year built_People_Before 1960', 'Year built_People_1960-2009',
       'Year built_Homes_Before 1960', 'Year built_Homes_1960-2009',
       'Year built_People_2010-9999', 'Year built_Homes_2010-9999',
       'Year built_Total people', 'Year built_Total homes'],
      dtype='object', length=146)


In [53]:
# Remove duplicate common columns
for year in years:
    df = combined_dfs[year]
    
    # Keep only the first occurrence of each common column
    cols_to_drop = [col for col in df.columns 
                    if any(col.startswith(f'{common}_') 
                          for common in ['ValgstedId', 'KredsNr', 'StorKredsNr', 'LandsdelsNr'])]
    df = df.drop(columns=cols_to_drop)
    combined_dfs[year] = df

df_2009, df_2013, df_2017, df_2021 = [combined_dfs[year] for year in years]

In [54]:
print(f"\n{'='*80}\n  2009\n")
print(df_2009.info(verbose=True, show_counts=True))
print(f"\n{'='*80}\n  2013\n")
print(df_2013.info(verbose=True, show_counts=True))
print(f"\n{'='*80}\n  2017\n")
print(df_2017.info(verbose=True, show_counts=True))
print(f"\n{'='*80}\n  2021\n")
print(df_2021.info(verbose=True, show_counts=True))


  2009

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58 entries, 0 to 57
Data columns (total 130 columns):
 #    Column                                              Non-Null Count  Dtype  
---   ------                                              --------------  -----  
 0    Gruppe                                              58 non-null     object 
 1    ValgstedId                                          58 non-null     object 
 2    KredsNr                                             58 non-null     object 
 3    StorKredsNr                                         58 non-null     object 
 4    LandsdelsNr                                         58 non-null     object 
 5    Socioeconomy_Self-employed                          58 non-null     float64
 6    Socioeconomy_Top executives                         58 non-null     float64
 7    Socioeconomy_Unemployed                             58 non-null     float64
 8    Socioeconomy_Total                                  58 non-nu

In [55]:
df_2021

,Gruppe,ValgstedId,KredsNr,StorKredsNr,LandsdelsNr,Socioeconomy_Self-employed,Socioeconomy_Top executives,Socioeconomy_Unemployed,Socioeconomy_Total,Sex_Male,...,Car_Households with 1 car,Car_Households with 2+ cars,Year built_People_Before 1960,Year built_People_1960-2009,Year built_Homes_Before 1960,Year built_Homes_1960-2009,Year built_People_2010-9999,Year built_Homes_2010-9999,Year built_Total people,Year built_Total homes
0,101001,101001,1,1,1,563.0,383.0,305.0,15540.0,7331.0,...,956,2032,13417.0,2055.0,6887.0,1117.0,90.0,40.0,15562.0,8044.0
1,101002,101002,1,1,1,357.0,221.0,137.0,8366.0,3908.0,...,658,1256,6994.0,1235.0,3418.0,809.0,149.0,72.0,8378.0,4299.0
2,101003,101003,1,1,1,632.0,401.0,277.0,15838.0,7640.0,...,807,2229,14942.0,904.0,7041.0,586.0,0.0,0.0,15846.0,7627.0
3,101005,101005,1,1,1,396.0,230.0,312.0,15817.0,7648.0,...,1057,1958,11810.0,3873.0,6119.0,2167.0,171.0,90.0,15854.0,8376.0
4,101006,101006,1,1,1,410.0,246.0,302.0,13473.0,6527.0,...,954,1802,10028.0,3240.0,5283.0,1737.0,219.0,85.0,13487.0,7105.0
5,101007,101007,2,1,1,377.0,202.0,293.0,14229.0,6970.0,...,856,2084,8275.0,5259.0,3950.0,2852.0,731.0,311.0,14265.0,7113.0
6,101008,101010;101008,2,1,1,1023.0,836.0,782.0,33823.0,16468.0,...,2049,4675,19096.0,11105.0,9939.0,5533.0,3442.0,1289.0,33643.0,16761.0
7,101009,101009,2,1,1,441.0,342.0,365.0,19097.0,15617.0,...,1193,4259,4452.0,13646.0,1659.0,6621.0,12964.0,5336.0,31062.0,13616.0
8,101011,101011,3,1,1,1062.0,520.0,381.0,20738.0,10914.0,...,916,1838,19560.0,690.0,7725.0,376.0,33.0,22.0,20283.0,8123.0
9,101012,101012,3,1,1,479.0,341.0,154.0,9173.0,4431.0,...,518,1565,9040.0,136.0,4033.0,95.0,0.0,0.0,9176.0,4128.0


## Drop columns

Now, we want to drop some columns:

- All 'Year built_[...]' columns for years 2009, 2013, 2017
- All columns with 0 non-null entries.
- All 'Education_[...]' columns that end with 'Medium higher education', 'Short higher education', 'Vocational upper secondary', 'General upper secondary' (keep all others).
- All 'Housing type_[...]' columns that don't end with 'Detached house', 'Terraced house' or 'Apartment buildings'

In [56]:
for year in years:
    df = combined_dfs[year]
    cols_to_drop = []
    
    # 1. Drop 'Year built_[...]' columns for 2009, 2013, 2017 (but not 2021)
    if year != 2021:
        cols_to_drop.extend([col for col in df.columns if col.startswith('Year built_')])
    
    # 2. Drop columns with 0 non-null entries
    cols_to_drop.extend([col for col in df.columns if df[col].notna().sum() == 0])

    # 3. Drop specific Education columns for all years
    education_suffixes = [
        'Medium higher education',
        'Short higher education',
        'Vocational upper secondary',
        'General upper secondary'
    ]
    cols_to_drop.extend([col for col in df.columns 
                        if col.startswith('Education_') and 
                        any(col.endswith(suffix) for suffix in education_suffixes)])
    
    # 3. Drop specific Housing type columns for all years
    housing_suffixes = [
        'Student halls',
        'Holiday homes',
        'Other homes',
    ]
    cols_to_drop.extend([col for col in df.columns 
                        if col.startswith('Housing type_') and 
                        any(col.endswith(suffix) for suffix in housing_suffixes)])
    
    # Remove duplicates and drop
    cols_to_drop = list(set(cols_to_drop))
    df = df.drop(columns=cols_to_drop)
    combined_dfs[year] = df

    print(f"Dropped {len(cols_to_drop)} columns from {year}, remaining: {df.shape[1]}")

# Reassign the cleaned dataframes
df_2009, df_2013, df_2017, df_2021 = [combined_dfs[year] for year in years]


Dropped 39 columns from 2009, remaining: 91
Dropped 37 columns from 2013, remaining: 93
Dropped 37 columns from 2017, remaining: 93
Dropped 17 columns from 2021, remaining: 101


In [57]:
print(f"\n{'='*80}\n  2009\n")
print(df_2009.info(verbose=True, show_counts=True))
print(f"\n{'='*80}\n  2013\n")
print(df_2013.info(verbose=True, show_counts=True))
print(f"\n{'='*80}\n  2017\n")
print(df_2017.info(verbose=True, show_counts=True))
print(f"\n{'='*80}\n  2021\n")
print(df_2021.info(verbose=True, show_counts=True))


  2009

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58 entries, 0 to 57
Data columns (total 91 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   Gruppe                                              58 non-null     object 
 1   ValgstedId                                          58 non-null     object 
 2   KredsNr                                             58 non-null     object 
 3   StorKredsNr                                         58 non-null     object 
 4   LandsdelsNr                                         58 non-null     object 
 5   Socioeconomy_Self-employed                          58 non-null     float64
 6   Socioeconomy_Top executives                         58 non-null     float64
 7   Socioeconomy_Unemployed                             58 non-null     float64
 8   Socioeconomy_Total                                  58 non-null     float

Later, we might consider:

- In 'Housing type_[...]', combining 'detached' and 'terraced' into one feature.
- Dropping the 'Benefit type_Folkepension' feature.
- Removing or merging some columns in 'Citizenship_[...]' and 'Immigrants_[...]'. E.g. drop 'Not specified', merge 'Old EU countries' with 'New EU countries' and 'Other Europe' (or find out what the differences are)

Another thing to note is that for housing tenure, housing type and education, the column naming format was a tiny bit different for 2009 (and earlier years) than for the other years. Because of this, we might consider dropping those columns from 2009 - or we might assume that it's just a tiny difference in the column naming format and nothing else.

### Add and drop geography columns
I want to remove StorKredsNr and LandsdelsNr, but add 'Valgsted navn', 'Kreds navn' and 'Kommune navn'. Preferably, the first six columns will be 'Gruppe', 'ValgstedId', 'Valgsted navn' 'KredsNr', 'Kreds navn' and 'Kommune navn', in that order.

In [58]:
def add_geography_columns(df, geography_csv_path):
    """
    Add geography columns to dataframe by matching ValgstedId.
    
    Parameters:
    - df: DataFrame to add columns to
    - geography_csv_path: Path to Geography.csv
    
    Returns:
    - DataFrame with added geography columns
    """    
    # Read Geography.csv
    geo_df = pd.read_csv(geography_csv_path, sep=';', encoding='utf-8', quotechar='"')
    
    # Clean column names (remove any extra whitespace/quotes)
    geo_df.columns = geo_df.columns.str.strip().str.strip('"')

    # Filter for KommuneNr 101 or 147
    geo_df = geo_df[geo_df['KommuneNr'].astype(str).isin(['101', '147'])].copy()

    # Debug: print first few rows to see what we're working with
    print("Geography CSV columns:", geo_df.columns.tolist())
    print("First few Valgsted Ids:", geo_df['Valgsted Id'].head().tolist())

    # Create a dictionary for quick lookup
    geo_dict = {}
    for _, row in geo_df.iterrows():
        # Clean the ValgstedId (remove quotes, whitespace)
        valgsted_id = str(row['Valgsted Id']).strip().strip('"')
        geo_dict[valgsted_id] = {
            'Valgsted navn': str(row['Valgsted navn']).strip().strip('"'),
            'Kreds navn': str(row['Kreds navn']).strip().strip('"'),
            'Kommune navn': str(row['Kommune navn']).strip().strip('"')
        }
    
    print(f"Total entries in geo_dict: {len(geo_dict)}")
    print(f"Sample keys: {list(geo_dict.keys())[:5]}")
    
    # Create new columns
    df_result = df.copy()
    valgsted_navne = []
    kreds_navne = []
    kommune_navne = []
    
    for _, row in df_result.iterrows():
        valgsted_ids = str(row['ValgstedId']).split(';')
        
        # Collect names for each ValgstedId
        v_names = []
        k_names = []
        kom_names = []
        
        for vid in valgsted_ids:
            vid = vid.strip()
            if vid not in geo_dict:
                print(f"Missing ValgstedId: '{vid}' (length: {len(vid)})")
                print(f"Available keys starting with '101': {[k for k in geo_dict.keys() if k.startswith('101')][:10]}")
                raise ValueError(f"ValgstedId '{vid}' not found in Geography.csv")
            
            v_names.append(geo_dict[vid]['Valgsted navn'])
            k_names.append(geo_dict[vid]['Kreds navn'])
            kom_names.append(geo_dict[vid]['Kommune navn'])
        
        # Join with semicolons, remove duplicates while preserving order
        valgsted_navne.append(';'.join(v_names))
        
        # For Kreds and Kommune, use single value if all same, otherwise semicolon-separate
        if len(set(k_names)) == 1:
            kreds_navne.append(k_names[0])
        else:
            kreds_navne.append(';'.join(k_names))
            
        if len(set(kom_names)) == 1:
            kommune_navne.append(kom_names[0])
        else:
            kommune_navne.append(';'.join(kom_names))
    
    # Add the new columns
    df_result['Valgsted navn'] = valgsted_navne
    df_result['Kreds navn'] = kreds_navne
    df_result['Kommune navn'] = kommune_navne
    
    # Drop the columns we don't want
    df_result = df_result.drop(columns=['StorKredsNr', 'LandsdelsNr'])
    
    # Reorder columns to put the first 6 in the desired order
    first_cols = ['Gruppe', 'ValgstedId', 'Valgsted navn', 'KredsNr', 'Kreds navn', 'Kommune navn']
    other_cols = [col for col in df_result.columns if col not in first_cols]
    df_result = df_result[first_cols + other_cols]
    
    return df_result

geography_path = ROOT + 'raw-data/historical-elections/election-results/Geography.csv'

for year in years:
    combined_dfs[year] = add_geography_columns(combined_dfs[year], geography_path)
    print(f"Added geography columns to {year}")

# Reassign
df_2009, df_2013, df_2017, df_2021 = [combined_dfs[year] for year in years]

# Check the result
print("\nFirst few rows of df_2021:")
print(df_2021[['Gruppe', 'ValgstedId', 'Valgsted navn', 'KredsNr', 'Kreds navn', 'Kommune navn']].head(10))

Geography CSV columns: ['Valgsted Id', 'KommuneNr', 'Kreds Nr', 'Storkreds Nr', 'Landsdels Nr', 'Valgsted navn', 'Kommune navn', 'Kreds navn', 'Storkreds navn', 'Landsdels navn', 'Valgsted start', 'Valgsted stop']
First few Valgsted Ids: [101001, 101002, 101003, 101005, 101006]
Total entries in geo_dict: 61
Sample keys: ['101001', '101002', '101003', '101005', '101006']
Added geography columns to 2009
Geography CSV columns: ['Valgsted Id', 'KommuneNr', 'Kreds Nr', 'Storkreds Nr', 'Landsdels Nr', 'Valgsted navn', 'Kommune navn', 'Kreds navn', 'Storkreds navn', 'Landsdels navn', 'Valgsted start', 'Valgsted stop']
First few Valgsted Ids: [101001, 101002, 101003, 101005, 101006]
Total entries in geo_dict: 61
Sample keys: ['101001', '101002', '101003', '101005', '101006']
Added geography columns to 2013
Geography CSV columns: ['Valgsted Id', 'KommuneNr', 'Kreds Nr', 'Storkreds Nr', 'Landsdels Nr', 'Valgsted navn', 'Kommune navn', 'Kreds navn', 'Storkreds navn', 'Landsdels navn', 'Valgsted s

## Convert (some) floats to ints

In [59]:
def convert_floats_to_int(df):
    """Convert float columns to int if they have no fractional parts"""
    df_converted = df.copy()
    
    for col in df_converted.columns:
        # Check if column is float type
        if df_converted[col].dtype == 'float64':
            # Check if all non-null values have no fractional part
            non_null_values = df_converted[col].dropna()
            if len(non_null_values) > 0 and (non_null_values % 1 == 0).all():
                # Convert to Int64 (nullable integer type to preserve NaN if present)
                df_converted[col] = df_converted[col].astype('int64')
    
    return df_converted

for year in years:
    combined_dfs[year] = convert_floats_to_int(combined_dfs[year])
    print(f"Converted {year}")

print("\ndf_2021 dtypes before conversion:")
print(df_2021.dtypes.value_counts())

# Reassign
df_2009, df_2013, df_2017, df_2021 = [combined_dfs[year] for year in years]

# Check the result
print("\ndf_2021 dtypes after conversion:")
print(df_2021.dtypes.value_counts())

Converted 2009
Converted 2013
Converted 2017
Converted 2021

df_2021 dtypes before conversion:
float64    83
int64      13
object      6
Name: count, dtype: int64

df_2021 dtypes after conversion:
int64     96
object     6
Name: count, dtype: int64


## Check total columns

In [60]:
# Check if 'total' columns are similar for the same polling area

total_cols = [
    'Gruppe',
    'Socioeconomy_Total',
    'Citizenship_Total',
    'Education_Total',
    'Housing tenure_Total people',
    'Housing type_Total people',
    'Benefit type_Antal personer i alt',
    'Year built_Total people'
]

total_df = df_2021[total_cols].copy()
total_df.head(10)

,Gruppe,Socioeconomy_Total,Citizenship_Total,Education_Total,Housing tenure_Total people,Housing type_Total people,Benefit type_Antal personer i alt,Year built_Total people
0,101001,15540,15562,13042,11438,15562,15562,15562
1,101002,8366,8378,6982,7017,8378,8378,8378
2,101003,15838,15846,12819,12193,15846,15846,15846
3,101005,15817,15854,13448,12350,15854,15854,15854
4,101006,13473,13487,11252,11023,13487,13487,13487
5,101007,14229,14265,11588,12900,14265,14265,14265
6,101008,33823,33643,27862,28022,33643,33643,33643
7,101009,19097,31062,24375,19813,31062,31062,31062
8,101011,20738,20283,17987,10167,20283,20283,20283
9,101012,9173,9176,7495,6444,9176,9176,9176


In [61]:
# They LOOK very similar, but let's see some statistics

numeric_totals = total_df.drop(columns=['Gruppe'])

total_df['mean_across_totals'] = numeric_totals.mean(axis=1)
total_df['median_across_totals'] = numeric_totals.median(axis=1)
total_df['mode_across_totals'] = numeric_totals.mode(axis=1)
total_df['std_across_totals'] = numeric_totals.std(axis=1)
total_df['min_across_totals'] = numeric_totals.min(axis=1)
total_df['max_across_totals'] = numeric_totals.max(axis=1)
total_df['range_across_totals'] = total_df['max_across_totals'] - total_df['min_across_totals']
total_df['cv_across_totals'] = (numeric_totals.std(axis=1) / numeric_totals.mean(axis=1)) * 100
# CV is the Coefficient of Variation (CV) - the relative variation as a percentage of the mean.

In [62]:
# Show rows with the most variation
total_df.nlargest(10, 'cv_across_totals')

,Gruppe,Socioeconomy_Total,Citizenship_Total,Education_Total,Housing tenure_Total people,Housing type_Total people,Benefit type_Antal personer i alt,Year built_Total people,mean_across_totals,median_across_totals,mode_across_totals,std_across_totals,min_across_totals,max_across_totals,range_across_totals,cv_across_totals
11,101014,28789,16714,14963,10052,16714,16714,16714,17237.142857,16714.0,16714,5647.095431,10052,28789,18737,32.761203
47,101055,15477,15656,12611,5782,15656,15656,15656,13784.857143,15656.0,15656,3703.415380,5782,15656,9874,26.865823
7,101009,19097,31062,24375,19813,31062,31062,31062,26790.428571,31062.0,31062,5578.125995,19097,31062,11965,20.821339
8,101011,20738,20283,17987,10167,20283,20283,20283,18574.857143,20283.0,20283,3816.364606,10167,20738,10571,20.545862
55,147005,14555,14285,12148,7414,14285,14285,14285,13036.714286,14285.0,14285,2612.240080,7414,14555,7141,20.037565
40,101048,11089,11095,9038,6776,11095,11095,11095,10183.285714,11095.0,11095,1686.539130,6776,11095,4319,16.561836
46,101054,4284,4468,4012,2678,4468,4468,4468,4120.857143,4468.0,4468,658.576097,2678,4468,1790,15.981532
25,101030,13679,13733,12058,9134,13733,13733,13733,12829.000000,13733.0,13733,1743.503656,9134,13733,4599,13.590332
9,101012,9173,9176,7495,6444,9176,9176,9176,8545.142857,9176.0,9176,1118.311887,6444,9176,2732,13.087106
53,147003,15733,15739,12867,11112,15739,15739,15739,14666.857143,15739.0,15739,1897.854791,11112,15739,4627,12.939751


The reason why 'Socioeconomy_Total' is so much higher than the other totals for 'Gruppe' == 101014, is probably that that 'Gruppe' contains two polling areas in the socioeconomy csv (101014 and 101057), but in the other csv files, it only contains 101014 (itself). That is kind of weird.

It looks like the totals for 'Citizenship_Total', 'Housing type_Total people', 'Benefit type_Antal personer i alt' and 'Year built_Total people' are always the same. But let's check it.

In [63]:
# Check if specific columns are always equal

cols_to_check = [
    'Citizenship_Total',
    'Housing type_Total people',
    'Benefit type_Antal personer i alt',
    'Year built_Total people'
]

# Check if all values in each row are equal
all_equal = (total_df[cols_to_check].nunique(axis=1) == 1).all()
print(f"Are these columns always equal across all rows? {all_equal}")

# See which rows (if any) have different values
rows_with_differences = total_df[total_df[cols_to_check].nunique(axis=1) > 1]
print(f"\nRows where values differ: {len(rows_with_differences)}")
if len(rows_with_differences) > 0:
    print(rows_with_differences[['Gruppe'] + cols_to_check])

# Verify by looking at a sample
print("\nSample of the columns:")
print(total_df[['Gruppe'] + cols_to_check].head(10))

Are these columns always equal across all rows? True

Rows where values differ: 0

Sample of the columns:
   Gruppe  Citizenship_Total  Housing type_Total people  \
0  101001              15562                      15562   
1  101002               8378                       8378   
2  101003              15846                      15846   
3  101005              15854                      15854   
4  101006              13487                      13487   
5  101007              14265                      14265   
6  101008              33643                      33643   
7  101009              31062                      31062   
8  101011              20283                      20283   
9  101012               9176                       9176   

   Benefit type_Antal personer i alt  Year built_Total people  
0                              15562                    15562  
1                               8378                     8378  
2                              15846                

Since they are always equal, I think we should set that value to 'Population' in a new column, delete the totals that were equal, and keep the others.

In [64]:
# Add 'Population' column and remove the redundant total columns
for year in years:
    df = combined_dfs[year]
    
    # Add 'Population' column with values from 'Citizenship_Total'
    df['Population'] = df['Citizenship_Total']
    
    # Drop the 4 redundant total columns
    cols_to_drop = [
        'Citizenship_Total',
        'Housing type_Total people',
        'Benefit type_Antal personer i alt',
        'Year built_Total people'
    ]
    if year == 2021:
        df = df.drop(columns=cols_to_drop)
    else:
        df = df.drop(columns=cols_to_drop[:3]) # Year built_ cols already dropped for other years
    combined_dfs[year] = df
    print(f"Updated {year}: {df.shape}")

df_2009, df_2013, df_2017, df_2021 = [combined_dfs[year] for year in years]

Updated 2009: (58, 90)
Updated 2013: (58, 92)
Updated 2017: (58, 92)
Updated 2021: (58, 99)


When calculating relative numbers for:
        'Citizenship_
        'Housing type_People_
        'Benefit type_
        'Year built_
        'Sex_
        'Age_'
        'Crime_'
Use 'Population' as the total to get the relative

Socioeconomy_
- Use Socioeconomy_Total 

Immigrants_
- Use Immigrants_Total

Education_
- Use Education_Total

Housing tenure_People_
- Use Housing tenure_Total people

Housing tenure_Homes_
- Use Housing tenure_Total homes

Housing type_Homes
- Use Housing type_Total homes

Car_
- Calculate total households from all three car columns

Income_Median household income
- Leave as is (already relative)

Income_Gross income top20pct
- Don't know right now :D maybe use Population

## Reorder and save to csv

In [65]:
column_order_prefixes = [
    'Sex_',
    'Age_',
    'Citizenship_',
    'Income_',
    'Car_',
    'Crime_',
    'Benefit type_',
    'Housing type_',
    'Year built_',
    'Housing tenure_',
    'Education_',
    'Immigrants_',
    'Socioeconomy_'
]

for year in years:
    df = combined_dfs[year]
    
    # Start with the first 7 columns
    first_cols = ['Gruppe', 'ValgstedId', 'Valgsted navn', 'KredsNr', 'Kreds navn', 'Kommune navn', 'Population']
    ordered_cols = first_cols.copy()
    
    # Add columns in the specified prefix order
    for prefix in column_order_prefixes:
        matching_cols = [col for col in df.columns if col.startswith(prefix) and col not in ordered_cols]
        ordered_cols.extend(sorted(matching_cols))  # Sort alphabetically within each group
    
    # Add any remaining columns that weren't caught (shouldn't be any, but just in case)
    remaining_cols = [col for col in df.columns if col not in ordered_cols]
    if remaining_cols:
        print(f"Warning: {year} has unexpected columns: {remaining_cols}")
        ordered_cols.extend(remaining_cols)
    
    # Reorder the dataframe
    df = df[ordered_cols]
    combined_dfs[year] = df
    print(f"Reordered {year}: {df.shape}")

# Reassign
df_2009, df_2013, df_2017, df_2021 = [combined_dfs[year] for year in years]

# Save to CSVs
for year in years:
    output_path = f'{OUTPUT_PATH}absolute_{year}.csv'
    combined_dfs[year].to_csv(output_path, index=False)
    print(f"Saved {output_path}")

# Verify the column order
print("\nColumn order for df_2021:")
for i, col in enumerate(df_2021.columns[:20], 1):
    print(f"{i}. {col}")


Reordered 2009: (58, 90)
Reordered 2013: (58, 92)
Reordered 2017: (58, 92)
Reordered 2021: (58, 99)
Saved ../../../processed-data/demographics/absolute_2009.csv
Saved ../../../processed-data/demographics/absolute_2013.csv
Saved ../../../processed-data/demographics/absolute_2017.csv
Saved ../../../processed-data/demographics/absolute_2021.csv

Column order for df_2021:
1. Gruppe
2. ValgstedId
3. Valgsted navn
4. KredsNr
5. Kreds navn
6. Kommune navn
7. Population
8. Sex_Female
9. Sex_Male
10. Age_0-4 years
11. Age_10-14 years
12. Age_15-17 years
13. Age_18-19 years
14. Age_20-24 years
15. Age_25-29 years
16. Age_30-34 years
17. Age_35-39 years
18. Age_40-44 years
19. Age_45-49 years
20. Age_5-9 years


And just like that, about 13,000 columns spread across 16 CSV files became less than 400 columns spread across 4 CSV files :-)